backend/vagrant.pm

```diff
@@ -0,0 +1,244 @@
+# Copyright © 2021 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package backend::vagrant;
+
+use strict;
+use warnings;
+
+use base 'backend::virt';
+use bmwqemu;
+use testapi ();
+
+use Mojo::Base -signatures;
+use Mojo::File;
+use Cwd qw(abs_path);
+use File::Temp;
+use File::chdir;
+use IPC::Run;
+
+sub new ($class) {
+    my $self = $class->SUPER::new;
+    my $vars = \%bmwqemu::vars;
+
+    $self->{vagrant_cwd} = File::Temp->newdir();
+    $self->{up_timeout}  = $vars->{VAGRANT_UP_TIMEOUT} // 300;
+    $self->{provider}    = $vars->{VAGRANT_PROVIDER}   // die 'Need variable \'VAGRANT_PROVIDER\'';
+    $self->{box_name}    = $vars->{VAGRANT_BOX} || die 'Need variable \'VAGRANT_BOX\'';
+    $self->{box_url}     = $vars->{VAGRANT_BOX_URL};
+
+    if (substr($self->{box_name}, 0, 1) eq '/') {
+        my $asset_dir    = $vars->{ASSETDIR} // '/var/lib/openqa/share/factory/';
```

Comment:
I would prefer to not introduce openQA-isms into backends. Can you make that a required variable with the PREFIX `VAGRANT_`?

Reply:
Do you mean `ASSETDIR`? I don't necessarily disagree with you, but I imagine that that will result in pretty bad UX, because now as a test writer you'll have to specify `ASSETDIR` twice or know the default.

Reply:
Why twice?

Reply:
Well, this breaks in the general case, e.g. one might use `OPENQA_BASEDIR` to use some other prefix than `/var/lib`. Besides, in productions workers assets are cached and we don't want to rely on the `/var/lib/openqa/share` mount.

Reply:
You could try the same as for HDD and ISO where by default we look in the current directory, I think

Reply:
> Why twice?

Let's assume your worker uses a different `ASSETDIR`, then you'd have to set `ASSETDIR` and `OPENQA_ASSETDIR` to the same value. That's what I meant by twice.


> Well, this breaks in the general case, e.g. one might use `OPENQA_BASEDIR` to use some other prefix than `/var/lib`. Besides, in productions workers assets are cached and we don't want to rely on the `/var/lib/openqa/share` mount.

Oh bummer. How can I get the cached assets then? Look into cwd as @okurz suggested?

Reply:
well, which assets are we actually talking about? And how do they end up in any folder? who or which component would copy them there at which time?

Reply:
I'm talking about vagrant boxes that are pushed to e.g. o3 from OBS staging project. At the moment they are available as `ASSET_LIBVIRT` and `ASSET_VIRTUALBOX` (see e.g. https://openqa.opensuse.org/tests/1803591#settings)

Reply:
@okurz I've tried to implement something like you suggested during today's QE tools meeting. Did you have something like that in mind?